pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,680 @@
+		for _, dep := range s.dependencies.GetDependencies() {
+			if dep.GetType() == "olm.gvk" {
+				gvk := registry.GVKDependency{}
+				if err := json.Unmarshal([]byte(dep.Value), &gvk); err != nil {
+					return nil, fmt.Errorf("error parsing %s as GVK, %v", dep.Value, err)
+				}
+				apiServiceGVK := v1.GroupVersionKind{
+					Group:   gvk.Group,
+					Version: gvk.Version,
+					Kind:    gvk.Kind,
+				}
+				if apiDes, ok := s.apiServiceDescriptionMap[apiServiceGVK]; ok {
+					requiredAPIServices = append(requiredAPIServices, *apiDes)
+					delete(s.apiServiceDescriptionMap, apiServiceGVK)
+				}
+			}
+		}
+	}
+
+	var ownedAPIServices []v1alpha1.APIServiceDescription
+	services, err := s.getServices()
+	if err != nil && err != ObjetctTypeNotFoundInBundle {
+		return nil, err
+	}
+
+	if services != nil {
+		for _, service := range *services {
+			gvk := service.GroupVersionKind()
+			apiGVKN := v1.GroupVersionKind{
+				Group:   gvk.Group,
+				Version: gvk.Version,
+				Kind:    gvk.Kind,
+			}
+			if apiDes, ok := s.apiServiceDescriptionMap[apiGVKN]; ok {
+				ownedAPIServices = append(ownedAPIServices, *apiDes)
+				delete(s.apiServiceDescriptionMap, apiGVKN)
+			} else {
+				ownedAPIServices = append(ownedAPIServices, v1alpha1.APIServiceDescription{
+					Name:    service.GetName(),
+					Group:   apiGVKN.Group,
+					Version: apiGVKN.Version,
+					Kind:    apiGVKN.Kind,
+				})
+			}
+		}
+	}
+
+	// Gets every APIService included in the bundle and check its GVK against RegistryV2CSV to see if a descriptor is
+	// provided. Add the descriptor if it is provided, else record GVKN. Skip if descriptor is provided as a CRD.
+	apiServices, err := s.getAPIServices()
+	if err != nil && err != ObjetctTypeNotFoundInBundle {
+		return nil, err
+	}
+
+	if apiServices != nil {
+		providedApiMap := make(map[v1.GroupVersionKind]struct{})
+		for _, apiService := range *apiServices {
+			providedApiMap[v1.GroupVersionKind{Group: apiService.Spec.Group, Version: apiService.Spec.Version}] = struct{}{}
+		}
+		for _, apiDes := range s.apiServiceDescriptionMap {
+			gv := v1.GroupVersionKind{Group: apiDes.Group, Version: apiDes.Version}
+			if _, ok := providedApiMap[gv]; ok {
+				ownedAPIServices = append(ownedAPIServices, *apiDes)
+				delete(providedApiMap, gv)
+			}
+		}
+		if len(providedApiMap) == 0 {
+			s.apiServiceDescriptionMap = map[v1.GroupVersionKind]*v1alpha1.APIServiceDescription{}
+		}
+	}
+
+	// Everything in apiServiceDescriptionMap either has to be specified in dependency.yaml as required or provided in
+	// the bundle.
+	if len(s.customResourceDefinitionMap) > 0 {
+		return nil, fmt.Errorf("not all descriptors are matched with provided or required APIServices or Service, "+
+			"including: %v", s.customResourceDefinitionMap)
+	}
+
+	return &v1alpha1.APIServiceDefinitions{
+		Owned:    ownedAPIServices,
+		Required: requiredAPIServices,
+	}, nil
+}
+
+func (s *synthesize) getCRDs() (*[]extv1.CustomResourceDefinition, error) {
```

Comment:
are you sure ? https://play.golang.org/p/PsHB1PZ_Tg4

Reply:
Yes, a slice is larger than a pointer, but [internally it holds only length info, capacity info, and a **pointer** to the underlying array](https://blog.golang.org/slices-intro#TOC_4.). Rather than size (since both are effectively constant), the tradeoff here is the amount of dereferencing that needs to be done to use the slice. Passing around slice values is [common in the go standard library](https://golang.org/src/hash/maphash/maphash.go?s=3375:3418#L210) as well as well.

This is definitely not blocking though.

Reply:
I see, no matter how large the slice is, it is the same size. Nice! thank you!